transactron/utils/amaranth_ext/memory.py

```diff
@@ -0,0 +1,332 @@
+                mem = memory.Memory(
+                    shape=port.width, depth=self.depth, init=self.init, attrs=self.attrs, src_loc_at=self.src_loc
+                )
+                m.submodules += mem
+                physical_write_port = mem.write_port(granularity=write_port.granularity)
+                physical_read_port = mem.read_port(
+                    transparent_for=[physical_write_port] if write_port in port.transparent_for else []
+                )
+                m.d.comb += [
+                    physical_read_port.addr.eq(port.addr),
+                    port.data.eq(physical_read_port.data),
+                    physical_read_port.en.eq(port.en),
+                    physical_write_port.addr.eq(write_port.addr),
+                    physical_write_port.data.eq(write_port.data),
+                    physical_write_port.en.eq(write_port.en),
+                ]
+
+        return m
+
+
+class MultiportXORMemory(BaseMultiportMemory):
+    """Multiport memory based on xor.
+
+    Multiple read and write ports can be requested. Memory is built of
+    (number of write ports) * (number of write ports - 1 + number of read ports) single port
+    memory blocks. XOR is used to enable writing multiple values in one cycle and reading correct values.
+    Writing two different values to the same memory address in one cycle has undefined behavior.
+
+    """
+
+    def elaborate(self, platform):
+        m = Module()
+
+        self._frozen = True
+
+        addr_width = bits_for(self.depth - 1)
+
+        write_xors: "list[Value]" = [Signal(self.shape, name="write_xor") for _ in self.write_ports]
+        read_xors: "list[Value]" = [Signal(self.shape, name="read_xor") for _ in self.read_ports]
+
+        write_regs_addr = [Signal(addr_width) for _ in self.write_ports]
+        write_regs_data = [Signal(self.shape) for _ in self.write_ports]
+        read_en_bypass = [Signal() for _ in self.read_ports]
+
+        for index, write_port in enumerate(self.write_ports):
+            if write_port is None:
+                raise ValueError("Found None in write ports")
+
+            index_passed_by = False
+            m.d.sync += [write_regs_data[index].eq(write_port.data), write_regs_addr[index].eq(write_port.addr)]
+            write_xors[index] ^= write_regs_data[index]
+            for i in range(len(self.write_ports) - 1):
+                mem = memory.Memory(
+                    shape=self.shape, depth=self.depth, init=[], attrs=self.attrs, src_loc_at=self.src_loc
+                )
+                mem_name = f"memory_{index}_{i}"
+                m.submodules[mem_name] = mem
+                physical_write_port = mem.write_port()
+                physical_read_port = mem.read_port(transparent_for=[physical_write_port])
+
+                if i == index:
+                    index_passed_by = True
+                idx = i + 1 if index_passed_by else i
```

Comment:
I believe this is equivalent, am I missing something?
```suggestion
                idx = i + 1 if i >= index else i
```

Reply:
This is equivalent and of course better